File: rxhttp/src/main/java/rxhttp/wrapper/parse/DownloadParser.java
package rxhttp.wrapper.parse;


import java.io.IOException;

import io.reactivex.annotations.NonNull;
import okhttp3.Response;
import okhttp3.ResponseBody;
import rxhttp.wrapper.exception.ExceptionHelper;
import rxhttp.wrapper.utils.IOUtil;

/**
 * 文件下载解析器
 * User: ljx
 * Date: 2018/10/23
 * Time: 13:49
 */
public class DownloadParser implements Parser<String> {
    private String mDestPath;

    /**
     * @param destPath 目标路径
     */
    public DownloadParser(@NonNull String destPath) {
        mDestPath = destPath;
    }

    /**
     * @param response Http请求执行结果
     * @return 下载成功后的文件路径
     * @throws IOException 网络异常等,RxJava的观察者会捕获此异常
     */
    @Override
    public String onParse(Response response) throws IOException {
        ResponseBody body = ExceptionHelper.throwIfFatal(response);
        boolean append = response.header("Content-Range") != null;
        IOUtil.write(body.byteStream(), mDestPath, append);//将输入流写出到文件
        return mDestPath;
    }
}

File: rxhttp/src/main/java/rxhttp/wrapper/parse/DownloadParser.kt
package rxhttp.wrapper.parse

import okhttp3.Response
import rxhttp.wrapper.OkHttpCompat
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

/**
 * File downloader
 * User: ljx
 * Date: 2018/10/23
 * Time: 13:49
 */
class DownloadParser(
    private var localPath: String,
) : StreamParser() {

    override fun onParse(response: Response): String {
        super.onParse(response)
        return localPath
    }

    override fun getOutputStream(response: Response): OutputStream {
        localPath = localPath.replaceSuffix(response)
        //创建文件
        val dstFile = File(localPath).apply {
            val parentFile = parentFile
            if (!parentFile.exists() && !parentFile.mkdirs()) {
                throw IOException("Directory $parentFile create fail")
            }
        }
        val append = OkHttpCompat.header(response, "Content-Range") != null
        return FileOutputStream(dstFile, append)
    }
}

private fun String.replaceSuffix(response: Response): String {
    return if (endsWith("/%s", true)
        || endsWith("/%1\$s", true)) {
        format(OkHttpCompat.pathSegments(response).last())
    } else {
        this
    }
}